core/pbcc_subscribe_v2.c

```diff
@@ -0,0 +1,259 @@
+    }
+    else {
+        PUBNUB_LOG_ERROR(
+            "No message array subscribe V2 response found, error=%d\n", jpresult);
+        return PNR_FORMAT_ERROR;
+    }
+
+    return PNR_OK;
+}
+
+
+struct pubnub_v2_message pbcc_get_msg_v2(struct pbcc_context* p)
+{
+    enum pbjson_object_name_parse_result jpresult;
+    struct pbjson_elem                   el;
+    struct pbjson_elem                   found;
+    struct pubnub_v2_message             rslt;
+    char const*                          start;
+    char const*                          end;
+    char const*                          seeker;
+
+    memset(&rslt, 0, sizeof rslt);
+
+    if (p->msg_ofs >= p->msg_end) {
+        return rslt;
+    }
+    start = p->http_reply + p->msg_ofs;
+    if (*start != '{') {
+        PUBNUB_LOG_ERROR(
+            "Message subscribe V2 response is not a JSON object\n");
+        return rslt;
+    }
+    end    = p->http_reply + p->msg_end;
+    seeker = pbjson_find_end_complex(start, end);
+    if (seeker == end) {
+        PUBNUB_LOG_ERROR(
+            "Message subscribe V2 response has no end of JSON object\n");
+        return rslt;
+    }
+
+    p->msg_ofs = (unsigned)(seeker - p->http_reply + 2);
+    el.start   = start;
+    el.end     = seeker;
+
+    /* @todo We should iterate over elements of JSON message object,
+       instead of looking from the start, again and again.
+    */
+
+    jpresult = pbjson_get_object_value(&el, "d", &found);
+    if (jonmpOK == jpresult) {
+        rslt.payload.ptr  = (char*)found.start;
+        rslt.payload.size = found.end - found.start;
+    }
+    else {
+        PUBNUB_LOG_ERROR("pbcc=%p: No message payload in subscribe V2 response "
+                         "found, error=%d\n",
+                         p,
+                         jpresult);
+        return rslt;
+    }
+
+    jpresult = pbjson_get_object_value(&el, "c", &found);
+    if (jonmpOK == jpresult) {
+        rslt.channel.ptr  = (char*)found.start + 1;
+        rslt.channel.size = found.end - found.start - 2;
+    }
+    else {
+        PUBNUB_LOG_ERROR("pbcc=%p: No message channel in subscribe V2 response "
+                         "found, error=%d\n",
+                         p,
+                         jpresult);
+        return rslt;
+    }
+
+    jpresult = pbjson_get_object_value(&el, "e", &found);
+    if (jonmpOK == jpresult) {
+        if (pbjson_elem_equals_string(&found, "1")) {
+            rslt.is_signal = true;
+        }
+        else {
+            rslt.is_signal = false;
+        }
+    }
+    else {
+        rslt.is_signal = false;
+    }
+    
+    jpresult = pbjson_get_object_value(&el, "p", &found);
+    if (jonmpOK == jpresult) {
+        struct pbjson_elem titel;
+        if (jonmpOK == pbjson_get_object_value(&found, "t", &titel)) {
+            if ((*titel.start != '"') || (titel.end[-1] != '"')) {
+                PUBNUB_LOG_ERROR("Time token in response is not a string\n");
```

Comment:
Is this better to attempt to cast it to an `int` and use that value throughout?

Reply:
Not quite sure what you mean, but, if you want to cast string to int, you can't do that in C :)

Reply:
oh good point.